File: livedata-lesson/src/main/java/com/acme/recipes/viewmodel/RecipeViewModel.java
package com.acme.recipes.viewmodel;

import android.arch.core.util.Function;
import android.arch.lifecycle.LiveData;
import android.arch.lifecycle.Transformations;
import android.arch.lifecycle.ViewModel;
import android.arch.lifecycle.ViewModelProvider;

import com.acme.recipes.database.dao.RecipeDao;
import com.acme.recipes.database.entity.RecipeEntity;
import com.acme.recipes.model.Recipe;

import io.realm.Realm;

public class RecipeViewModel extends ViewModel {

    private final Realm database;
    private final RecipeDao dao;

    private LiveData<Recipe> recipe;

    public RecipeViewModel(String recipeId) {
        database = Realm.getDefaultInstance();
        dao = new RecipeDao(database);
        recipe = Transformations.map(dao.findByIdAsync(recipeId),
                new Function<RecipeEntity, Recipe>() {
            @Override
            public Recipe apply(RecipeEntity input) {
                return input;
            }
        });
    }

    public LiveData<Recipe> getRecipe() {
        return recipe;
    }

    @Override
    protected void onCleared() {
        database.close();
    }


    public static class Factory extends ViewModelProvider.NewInstanceFactory {

        private final String recipeId;

        public Factory(String recipeId) {
            this.recipeId = recipeId;
        }

        @Override
        public <T extends ViewModel> T create(Class<T> modelClass) {
            //noinspection unchecked
            return (T) new RecipeViewModel(recipeId);
        }
    }


}

File: app/src/main/java/com/acme/recipes/viewmodel/RecipeViewModel.kt
package com.acme.recipes.viewmodel

import android.arch.lifecycle.LiveData
import android.arch.lifecycle.Transformations
import android.arch.lifecycle.ViewModel
import android.arch.lifecycle.ViewModelProvider
import com.acme.recipes.database.dao.RecipeDao
import com.acme.recipes.database.util.recipeDao
import com.acme.recipes.model.Recipe
import io.realm.Realm

class RecipeViewModel(recipeId: String) : ViewModel() {

    private val database = Realm.getDefaultInstance()
    private val dao = database.recipeDao()

    val recipe: LiveData<Recipe>

    init {
        recipe = Transformations.map(dao.findByIdAsync(recipeId)) {
            // Map from LiveData<RecipeEntity> to LiveData<Recipe>
            input -> input
        }
    }

    override fun onCleared() {
        database.close()
    }


    class Factory(private val recipeId: String) : ViewModelProvider.NewInstanceFactory() {

        override fun <T : ViewModel> create(modelClass: Class<T>): T {

            @Suppress("UNCHECKED_CAST")
            return RecipeViewModel(recipeId) as T
        }
    }


}
